docs/Serial.md

```diff
@@ -56,86 +56,151 @@ e.g. after configuring a port for GPS enable the GPS feature.
- | FUNCTION_TELEMETRY_IBUS | 4096 |
- | FUNCTION_VTX_TRAMP | 8192 |
- | FUNCTION_RCDEVICE | 16384 |
- | FUNCTION_LIDAR_TF | 32768 |
- | FUNCTION_FRSKY_OSD | 65536 |
-
-
-
-### Baud Rates
-
-The allowable baud rates are as follows:
-
-| Identifier | Baud rate |
-| ---------- | --------- |
-| 0          | Auto      |
-| 1          | 9600      |
-| 2          | 19200     |
-| 3          | 38400     |
-| 4          | 57600     |
-| 5          | 115200    |
-| 6          | 230400    |
-| 7          | 250000    |
+The `serial` CLI command takes 6 arguments:
+```
+serial <serial port identifier> <serial port function> <msp baudrate> <sensor baudrate> <telemetry baudrate> <peripheral baudrate>
+```
 
+| Argument                       | Default | Source                   |
+| ------------------------------ | ------- | ------------------------ |
+| 1. Serial Port Identifier      |         | serialPortIdentifier_e   |
+| 2. Serial Port Function        |         | serialPortFunction_e     |
+| 3. Configuration/MSP baud rate | 115200  | baudrate_e               |
+| 4. Sensor Input baud rate      | 57600   | gpsbaudRate_e            |
+| 5. Telemetry baud rate         | AUTO    | baudRate_e               |
+| 6. Peripherals baud rate       | 115200  | baudrate_e               |
+
+### 1. Serial Port Identifier
+
+| Identifier                 | Value |
+| -------------------------- | ----- |
+| SERIAL_PORT_NONE           | -1    |
+| SERIAL_PORT_USART1         | 0     |
+| SERIAL_PORT_USART2         | 1     |
+| SERIAL_PORT_USART3         | 2     |
+| SERIAL_PORT_UART4          | 3     |
+| SERIAL_PORT_UART5          | 4     |
+| SERIAL_PORT_USART6         | 5     |
+| SERIAL_PORT_USART7         | 6     |
+| SERIAL_PORT_USART8         | 7     |
+| SERIAL_PORT_LPUART1        | 8     |
+| SERIAL_PORT_USB_VCP        | 20    |
+| SERIAL_PORT_SOFTSERIAL1    | 30    |
+| SERIAL_PORT_SOFTSERIAL2    | 31    |
+| SERIAL_PORT_IDENTIFIER_MAX | 31    |
+
+Note: `SERIAL_PORT_IDENTIFIER_MAX` = `SERIAL_PORT_SOFTSERIAL2`
+
+### 2. Serial Port Function
+
+ | Function                     | Value |
+ | ---------------------------- | ----- |
+ | FUNCTION_NONE                | 0     |
+ | FUNCTION_MSP                 | 1     |
+ | FUNCTION_GPS                 | 2     |
+ | FUNCTION_TELEMETRY_FRSKY_HUB | 4     |
+ | FUNCTION_TELEMETRY_HOTT      | 8     |
+ | FUNCTION_TELEMETRY_LTM       | 16    |
+ | FUNCTION_TELEMETRY_SMARTPORT | 32    |
+ | FUNCTION_RX_SERIAL           | 64    |
+ | FUNCTION_BLACKBOX            | 128   |
+ | FUNCTION_TELEMETRY_MAVLINK   | 512   |
+ | FUNCTION_ESC_SENSOR          | 1024  |
+ | FUNCTION_VTX_SMARTAUDIO      | 2048  |
+ | FUNCTION_TELEMETRY_IBUS      | 4096  |
+ | FUNCTION_VTX_TRAMP           | 8192  |
+ | FUNCTION_RCDEVICE            | 16384 |
+ | FUNCTION_LIDAR_TF            | 32768 |
+ | FUNCTION_FRSKY_OSD           | 65536 |
+
+Note: `FUNCTION_FRSKY_OSD` = `(1<<16)` requires 17 bits.
+
+### 3. Configuration/MSP Baud Rates
```

Comment:
Have a look at https://github.com/betaflight/betaflight/blob/d3ad63a50bbb48c39a1a4d4d6ceaff4a95c9fba4/src/main/cli/cli.c#L1334-L1359 - the value ranges you are listing do not match the actual ranges in the firmware.

Reply:
I looked at the code in gps.c including the headers. Looks like the values are constraint by the table I specified.
Also GUI is more restrictive than CLI for setting values which seems more generic. So need some guidance here.

GPS:
https://github.com/betaflight/betaflight/blob/master/src/main/io/gps.h#L58-L64
https://github.com/betaflight/betaflight/blob/master/src/main/io/gps.c#L108-L123

Blackbox:
https://github.com/betaflight/betaflight/blob/master/src/main/blackbox/blackbox_io.c#L326-L337 

The msp and telemetry baudrates I am not sure (I got those from configurator).
https://github.com/betaflight/betaflight-configurator/blob/master/src/js/tabs/ports.js#L64-L107

Sorry, I don't know how the code expands (cannot insert permalink) within this comment.

According to above code references which values would you want me to document?

Reply:
I think if there are discrepancies between firmware and configurator as to the allowable ranges for these parameters, we should first clean up these discrepancies, and then document what the actual ranges are - and most likely it is the firmware that is correct, and it is just a case of firmware changes that were never ported into configurator.

On closer look at configurator, the only discrepancy that I could find is
https://github.com/betaflight/betaflight-configurator/blob/74820fae37269af739490a3bfcea1ba3265ca67c/src/js/tabs/ports.js#L97

With blackbox being a TX-only protocol, `AUTO` for baud rate selection does not make the slightest sense, as the firmware will have no way to 'detect' the baud rate of the peripheral. Turns out 'AUTO' is only meant to be used for non-blackbox peripherals (in a super clumsy way): https://github.com/betaflight/betaflight-configurator/blob/74820fae37269af739490a3bfcea1ba3265ca67c/src/js/tabs/ports.js#L212-L217

The missing baud rates for MSP are added here: https://github.com/betaflight/betaflight-configurator/blob/74820fae37269af739490a3bfcea1ba3265ca67c/src/js/tabs/ports.js#L74-L76

Reply:
Thanks missed the concat updating the msp array.